### PR TITLE
Refactored Plugin related code

### PR DESCRIPTION
**Pull request by Febbe**
- use std::unique_ptr
 - removed XournalType.h
 - fixed many clang-tidy warnings.
 - consistently use fs::path
 - mark functions override instead of virtual

### CODE REVIEW

**Comment by rolandlo** [comment]
I was just looking over the code. Could you tell me, what you changed with your last commit/force-push? 
One remark: I needed to add `#include <memory>` to `PluginDialogEntry.h` to make it compile with gcc 9.2.1

**Comment by Febbe** [comment]
> I was just looking over the code. Could you tell me, what you changed with your last commit/force-push?
> One remark: I needed to add `#include <memory>` to `PluginDialogEntry.h` to make it compile with gcc 9.2.1

I just added the #include <memory> in 2 files and fixed the clang-format. I normally use clang-12 to compile my cpp code. And clang is very friendly when it comes to missing header files. So it always compiles on my computer :D.

**Comment by Febbe** [comment]
First of all I need feedback if all plugins still load and work as expected. I also need feedback if plugins in special non-ASCII folder names still work on Windows. Unfortunately I can't test this very often because my Windows PC is rarely accessible.

**Comment by rolandlo** [comment]
> First of all I need feedback if all plugins still load and work as expected. I also need feedback if plugins in special non-ASCII folder names still work on Windows. Unfortunately I can't test this very often because my Windows PC is rarely accessible.

I have run your PR (compiled from source) from the terminal via `build/src/xournalpp`. All plugins in the `plugins` folder in the same `xournalpp` directory load as expected. To see filenames displayed lexically normal is certainly a plus. Otherwise all the plugins seem to work as before. 
One thing that bugs me is that accelerators like `<Release>z` don't seem to work. Firstly it is displayed like `Z` (as if there was no `<Release>` before the `z` key in the menu. Secondly pressing and releasing the `z` key does not execute the plugin. I don't know if I misunderstand something, but in the [GTK documentation](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html#gtk-accelerator-parse) it is written 

> The format looks like “\<Control\>a” or “\<Shift\>\<Alt\>F1” or “\<Release\>z” (the last one is for key release).

So I would expect it to work.

Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).

I will now check how your PR works on Windows.

**Comment by rolandlo** [comment]
Under windows there is something wrong with the terminal messages: 
grafik

**Comment by rolandlo** [comment]
Special folder names do not seem to work on Windows. I renamed the ColorCycle plugin to `àÄ@fix~d$$äö!è`, after which it was still correctly displayed in the Plugin Manager, but could not be loaded upon activating it and restarting Xournal++, see
grafik

**Comment by Febbe** [comment]
@rolandlo the terminal in windows don't support utf8 by default. Sometimes it isn't even ASCII/ANSI. To fix that all << and >> musst be replaced by "

**Comment by Febbe** [comment]
> One thing that bugs me is that accelerators like `<Release>z` don't seem to work. Firstly it is displayed like `Z` (as if there was no `<Release>` before the `z` key in the menu. Secondly pressing and releasing the `z` key does not execute the plugin. I don't know if I misunderstand something, but in the [GTK documentation](https://developer.gnome.org/gtk3/stable/gtk3-Keyboard-Accelerators.html#gtk-accelerator-parse) it is written

Has the issue been there before?  I only changed the path representation, the rest is code cleanup. I also don't quite understand what the problem is. What is the difference between accelerators like `<Release>z' and those that work.

**Comment by rolandlo** [comment]
> Has the issue been there before? I only changed the path representation, the rest is code cleanup. I also don't quite understand what the problem is. What is the difference between accelerators like `z' and those that work.

Yes, this issue has been there before. The idea of using accelerators like `<Release>z` is for temporary activation of some tool. For instance you want to switch to the eraser tool when pressing `z` and switch back to the previously used tool when releasing this key. Accelerators work, but the key release accelerators (that should act upon releasing the key) don't.

**Comment by Febbe** [comment]
> Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).

I think that's a great idea. There is already a user config folder in `~/.config/xournalpp` there should be another folder "plugins" in there, I can do this.

**Comment by Febbe** [comment]
> Yes, this issue has been there before. The idea of using accelerators like `<Release>z` is for temporary activation of some tool. For instance you want to switch to the eraser tool when pressing `z` and switch back to the previously used tool when releasing this key. Accelerators work, but the key release accelerators (that should act upon releasing the key) don't.

I didn't wrote the plugin code, so I don't know where the accelerators are handled. I can imagine, that the accelerators won't be handled, if they are already handled. This must be investigated. You can add this to #2354 by editing my post.

File: src/plugin/Plugin.cpp [hunk 1]
```diff
@@ -282,33 +203,33 @@ void Plugin::loadScript() {
...
     addPluginToLuaPath();
 
     // Run the loaded Lua script
-    if (lua_pcall(lua, 0, 0, 0) != LUA_OK) {
-        const char* errMsg = lua_tostring(lua, -1);
-        map<int, string> button;
-        button.insert(std::pair<int, string>(0, _("OK")));
+    if (lua_pcall(lua.get(), 0, 0, 0) != LUA_OK) {
+        const char* errMsg = lua_tostring(lua.get(), -1);
+        map<int, std::string> button;
```
Comment (Technius) [inline src/plugin/Plugin.cpp:231]: For consistency, this should also be `std::map`.

File: src/plugin/Plugin.cpp [hunk 2]
```diff
@@ -317,17 +238,14 @@ void Plugin::loadScript() {
...
+    lua_getglobal(lua.get(), fnc.c_str());
 
     // Run the function
-    if (lua_pcall(lua, 0, 0, 0)) {
-        const char* errMsg = lua_tostring(lua, -1);
-        map<int, string> button;
-        button.insert(std::pair<int, string>(0, _("OK")));
+    if (lua_pcall(lua.get(), 0, 0, 0)) {
+        const char* errMsg = lua_tostring(lua.get(), -1);
+        map<int, std::string> button;
```
Comment (Technius) [inline src/plugin/Plugin.cpp:247]: `std::map` here too.

File: src/plugin/Plugin.cpp [hunk 3]
```diff
@@ -16,66 +14,38 @@ extern "C" {
...
-/**
- * Get Plugin from lua engine
- */
-Plugin* Plugin::getPluginFromLua(lua_State* lua) {
+auto Plugin::getPluginFromLua(lua_State* lua) -> Plugin* {
     lua_getfield(lua, LUA_REGISTRYINDEX, "Xournalpp_Plugin");
 
     if (lua_islightuserdata(lua, -1)) {
-        Plugin* data = (Plugin*)lua_touserdata(lua, -1);
+        auto* data = (Plugin*)lua_touserdata(lua, -1);
```
[the remark this answers is not in the document]
Comment (Technius) [reply src/plugin/Plugin.cpp:32]: Or `static_cast`, whichever you think is more readable here (I have no opinion, they seem equally clear to me).

**Comment by Technius** [comment]
> > Another question: Could we add a user plugin folder? Currently plugins are only loaded from one specific folder (which is a system folder in case Xournal++ has been installed on the system, and which cannot be modified without administrator rights).
> 
> I think that's a great idea. There is already a user config folder in `~/.config/xournalpp` there should be another folder "plugins" in there, I can do this.

See #1155 for the relevant issue.

File: src/plugin/Plugin.cpp (hunk 2 above)
Comment (Febbe) [reply src/plugin/Plugin.cpp]: done, also added a todo in our metaissue

**Comment by Febbe** [comment]
> Special folder names do not seem to work on Windows. I renamed the ColorCycle plugin to `àÄ@fix~d$$äö!è`, after which it was still correctly displayed in the Plugin Manager, but could not be loaded upon activating it and restarting Xournal++, see
> grafik
> 
> **Edit** even with a simple folder name like `René` or `Ähnlichkeit` the plugin cannot be loaded under Windows.

It seems like, this is a dependency issue caused by mingw-64. If the path exist, fopen should be able to open path::string::c_str

**Pull request merged**